### DISCUSSION TITLE
Regarding TypeDecorator

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#7271] [Usage Questions]
TL;DR: Want to print out SQLA sql statements literally, even with types not natively handled by SQLA? Look at #THIS WORKS cases in the code below, and add stuff to LiteralDialect as you need...

Can someone make #THIS WILL FAIL work without having to declare the column in Foo to be of the UUIDTypeDecorator type?

This post may prove useful to anyone looking to debug arbitrary sql expressions by looking at the actual SQL that would be emitted, including parameter values. The original implementation of literalquery() is something I found somewhere online long ago, and I've forgotten the source. I'd like to credit the original author but I remember have to dig and dig to find it, and I don't have time right now to find it again. 

The documentation for SQLA stands to be improved in the section: Rendering Bound Parameters Inline and also in the TypeDecorator section. It's not at all clear from reading these sections whether these "decorators" are somehow registered in ways that causes them to be picked up automatically by SQLA merely by declaring the classes (i.e., merely by instantiating the classes as Python classes).

Obviously the term "Decorator" has a specific OO meaning, but the documentation gives the impression that SQLA's type system is picking up declared classes without explicitly tying those "decorated" declarations to ORM columns -- only to types that could be on any column. I.e., that simply creating a TypeDecorator subclass somehow registers it in such a way that it will be used  where the "#THIS WILL FAIL" comment below exists. Nor does reading the TypeDecorator documentation clarify how these decorators should be *used* in client code. The very use-case discussed in the "Rendering Bound Parameters" section is most useful if you don't have to go and change all your ORM declarations to use some specific decorated type just to debug your code (or someone else's).

Instead, the only way I could find to actually cause a decorator to be used is to declare a given column as being of the decorator class type! This seems counter to the intent of the section "Rendering Bound Parameters" because the implication is that users must then cause all their declared ORMs to use that decorator type. How can one possibly expect to render an sql statement built in a complex system using implemented decorator functions if you have to go and change all the columns to BE the decorator type?

I must be missing something!!!  Can someone please tell me if some simple tweak to the code below would cause the line at "#THIS WILL FAIL" to run absent what I do in the Foo2 ORM class declaration?

It's also really unfortunate that the SQLA documentation for dialects does not seem to indicate exactly how to refer to a particular dialect. I had to go read the source code to figure out that the default psycopg2 postgresql dialect is refered to in python as `sqlalchemy.dialects.postgresql.dialect`  ... those dialect tables in that section of the documentation should probably list the actual literal to use for each dialect!

```py
import sqlalchemy as sa
from sqlalchemy import Column, String, DateTime, DECIMAL, Boolean, ForeignKey, BIGINT, text, TypeDecorator
from sqlalchemy.dialects import postgresql
from sqlalchemy.ext.declarative import declarative_base
import sqlalchemy.orm
from sqlalchemy.engine.default import DefaultDialect
from sqlalchemy.sql.sqltypes import String, DateTime, NullType, Enum
from uuid import UUID
from sqlalchemy.dialects import postgresql
import traceback

# python2/3 compatible.
PY3 = str is not bytes
text = str if PY3 else unicode
int_type = int if PY3 else (int, long)
str_type = str if PY3 else (str, unicode)

class StringLiteral(String):
	"""Teach SA how to literalize various things."""
	def literal_processor(self, dialect):
		super_processor = super(StringLiteral, self).literal_processor(dialect)

		def process(value):
			if isinstance(value, int_type):
				return text(value)
			if not isinstance(value, str_type):
				value = text(value)
			result = super_processor(value)
			if isinstance(result, bytes):
				result = result.decode(dialect.encoding)
			return result
		return process

class LiteralDialect(DefaultDialect):
	colspecs = {
		# prevent various encoding explosions
		String: StringLiteral,
		# teach SA about how to literalize a datetime
		DateTime: StringLiteral,
		# don't format py2 long integers to NULL
		NullType: StringLiteral,
		postgresql.UUID: StringLiteral
	}

literal_dialect = LiteralDialect()

class UUIDTypeDecorator(TypeDecorator):
	impl = postgresql.UUID
	 
	def process_literal_param(self, value, dialect):
		return str(value)


def literalquery(statement,kwargs={},literal_binds=True,dialect=literal_dialect):
	"""NOTE: This is entirely insecure. DO NOT execute the resulting strings."""
	if isinstance(statement, sqlalchemy.orm.Query):
		statement = statement.statement
	return statement.params(kwargs).compile(
		dialect=dialect,
		compile_kwargs=({'literal_binds': True} if literal_binds else {})  ).string

#----------------------------------------------------------------------------------------------------

orm_class_registry = {}
ORM = declarative_base(metadata=sa.MetaData(),class_registry=orm_class_registry)
class Foo (ORM):
	__tablename__ = 'foo'
	id = Column(postgresql.UUID,primary_key=True)

f = Foo.__table__
sql = sa.select([f.c.id]).where(f.c.id == UUID(int=0))

#THIS WORKS
print(literalquery(sql))

try:
	# THIS WILL FAIL!!!
	print(literalquery(sql,dialect=postgresql.dialect()))
except Exception as e:
	traceback.print_exc()

#----------------------------------------------------------------------------------------------------

class Foo2 (ORM):
	__tablename__ = 'foo2'
	id = Column(UUIDTypeDecorator,primary_key=True)

f = Foo2.__table__
sql = sa.select([f.c.id]).where(f.c.id == UUID(int=0))

#THIS WORKS
print(literalquery(sql,dialect=postgresql.dialect()))
```

**Comment by CaselIT** [comment]
Hi,

I'm not sure what the question is.
The postgresql dialect does not have a literal processor for uuid, so it will fail if you try to compile with `literal_binds=True`.
If you provide a type decorator that implements a literal processor the postgresql compiler can use it to render the object as literal.

Make your literal dialect a subclass of the postgresql dialect and you should be set. You may probably want to add the default postgresql colspecs to it though: `colspecs = { **postgresql.dialect.colspecs, ...`

**Reply by aathan** [reply]
Say what now?
What does "provide a type decorator that implements a literal processor" mean? Can you be explicit regarding the actions behind the verb "provide"?

Let me make sure we all understand the goal here: There's a big complex system that someone is trying to debug. They don't control or even necessarily understand all parts of the system. They don't want to perturb the system too much, but they DO want to get useful debugging output at some random line somewhere in the codebase where a variable `sql` contains an SQLA statement that is not returning the data they expect.

SQLA expects the programmer to inject an entirely new dialect and cause it to be used "at the top level" for all its behaviors leading to where the `sql` is then again available to print out properly?!

Alternatively, SQLA expects the programmer to print `str(sql)` which could be some giant query with joins for all the relationship data and 50 interpolated variables? And that string is in a format that cannot be easily caused to contain reasonable repr()'s or str()'s of the variables?

I beg for mercy oh great SQLA gods.

**Reply by zzzeek** [reply]
can you tone the sarcasm down a bit?  we're old.   pretend we all work together and we're face to face.    thanks

**Reply by CaselIT** [reply]
> What does "provide a type decorator that implements a literal processor" mean?

this is a description of a `UUIDTypeDecorator` like class you provided in the example above...

Reading trough the sarcasm, it seem you want to turn on logging. You can read how to do that here https://docs.sqlalchemy.org/en/14/core/engines.html#dbengine-logging

**Reply by aathan** [reply]
Absolutely. It looks like we were responding concurrently and I totally appreciate your responses and attention. I didn't intend to be sarcastic so much as describe an actual situation -- and I really did mean to expose where and how the existing documentation may be less useful to someone first encountering these issues in SQLA vs experts such as yourselves.

Yes, I figured out how to turn on logging as a poor alternative to this. This can unfortunately lead to such volumes of output, and slow the system down enough, to be less useful that I'd hoped.


In other words, I was actually being serious about the need to elaborate on the word "provide" -- because what the TypeDecorator documentation is not clear about is how to "provide" the newly implemented typedecorator to the system.

**Reply by CaselIT** [reply]
Regarding type decorator, I believe sqlalchemy have had no previous complaint regarding how would they be need to be used.
They are usually used to customize the behavior of types, and would not be my first choice regarding literal binds.

**Reply by aathan** [reply]
I think it's a path dependent issue caused by coming into the prior docs at that point, combined with the word "decorator" used in the "TypeDecorator" class name. In my experience decorators are added to existing classes (often dynamically) vs being used directly on their own. Therefore, the previous example left me wondering whether that's what was happening.

The new doc that was posted today which states "The above datatype needs to be used either explicitly within the model or locally within the statement using type_coerce(), such as" provides exactly the clarification I'm talking about. Thank you.

I'll see if I can find the source for these docs so that in the future, I can provide concrete changes for your consideration, rather than only "complain." My goal was to improve the docs by illustrating where the confusion originates.

**Reply by CaselIT** [reply]
I guess TypeDecorator it's called decorator since it augments an existing type, it's not creating a new one, like UserDefinedType

Feel free to provide changes to the docs.

The non api documentation is here https://github.com/sqlalchemy/sqlalchemy/tree/main/doc/build in rst format that is compiled with sphinx. The api docs are instead inline with the code. Usually searching for a small section of a phase will point you where it's written

**Comment by zzzeek** [comment]
> How can one possibly expect to render an sql statement built in a complex system using implemented decorator functions if you have to go and change all the columns to BE the decorator type?

well the main issue is that we really don't want people doing this at all.   SQLAlchemy wants to pass all the literal values onto the DBAPI, which is the software that knows how to correctly bind these values to a statement, either via API calls to a native system that accepts the parameters separately, or using carefully developed / tested / maintained string formatting functions that are local to that DBAPI.  SQLAlchemy doesn't want to duplicate that work, or support it, or get CVEs when vulnerabilities are found.    

The SQL construct you have supports a simple stringify by just calling str(stmt); this will use a special stringify-only dialect that will in fact print out your UUID value, as you note in your "# THIS WORKS!!" section.

for a stringify with parameters inlined for PostgreSQL, I would recommend you use the psycopg2 [mogrify](https://www.psycopg.org/docs/cursor.html#cursor.mogrify) method using a raw cursor.  for raw cursor usage see https://docs.sqlalchemy.org/en/14/core/connections.html#working-with-the-dbapi-cursor-directly .

**Reply by zzzeek** [reply]
btw here's how you get any dialect:

```py
e = create_engine("postgresql+psycopg2://")
the_dialect = e.dialect
```

**Reply by aathan** [reply]
I do understand and appreciate these valid concerns.

I would also argue that SQLA is already a fiendishly complex system (as most OO to relational layers are) that can be incredibly opaque for the uninitiated AND that debugging a SQL query is ultimately *the most frequent and desirable act* that programers must perform. I imagine 99% of programmers who don't get the data they expect want to directly interface with the underlying database, run some queries, and figure out what about the relational statement is wrong.

To your point, the problem could be the rendering of literal values or anything in between, so ideally, the programmer would then have access to the actual SQL emitted to the DB by the underlying DBAPI. But how?! How often have I said to myself "I just wish SQLA would get out of my way" despite how useful so much of it is.

Perhaps SQLA should at the very least provide methods that call the right exposed functions of the DBAPI to perform the equivalent of a useful `print(literalquery(sql))` for each supported DBAPI! I say that with the guile of a noob when it comes to the issues. Regardless, it's what I'd wish existed and what I wish was documented instead of what you find when you google reasonable search terms and end up at "Rendering Bound Parameters."  <--- that, by the way, is documentation that sends you down entirely the wrong rabbit hole, obviously.

As far as the "simple stringify" I'd like to better understand what you suggest to be a useful alternative to the complex solution I posted, or to the much shorter, but equally difficult to grok from the SQLA documentation, solution posted in your next response below.  For me `str(sql)` outputs stuff like `...WHERE foo.id = :id_1`

That format may seem useful, but in a real system, particularly with join fetches of relationships, the output ends up an ugly giant mess that you cannot easily interpolate values into, crucially, is nowhere close to what could be actually run on a SQL commandline.

Perhaps another idea would be to make the `str(sql)` output something you COULD pass to a pythong string formatter along with a dict of the values??

**Reply by CaselIT** [reply]
> To your point, the problem could be the rendering of literal values or anything in between, so ideally, the programmer would then have access to the actual SQL emitted to the DB by the underlying DBAPI

Sqlalchemy can only provide what it passes to the dbapi, it has no control at what the dbapi then does with it. Also if you want that you are probably better off at looking at the database logging stream, so you get all the sql that is executed.

> Perhaps SQLA should at the very least provide methods that call the right exposed functions of the DBAPI to perform the equivalent of a useful `print(literalquery(sql))` for each supported DBAPI!

This is not feasible, since not all dialect have something like this (like the one that use server side binding) or if the have them they don't expose them.

> Perhaps another idea would be to make the `str(sql)` output something you COULD pass to a pythong string formatter along with a dict of the values??

This would be a significant breaking change. Also, as Mike mentioned, sqlalchemy does not wish to encourage stratification of the sql string for various and well documented security concerns.

**Reply by aathan** [reply]
I often end up identifying a repeatable pattern and associated set of inputs to the system that generate buggy behavior, and then (especially in dynamic languages like Python) can zone in on the problem statements and write a logical expression that filters the otherwise voluminous activity to the case I'm trying to fix. I'm not sure if that's common, but that's what I've frequently found myself to be doing with SQLA systems I work on.

It's at that exact point that I then want to see what's being done by the underlying frameworks. I don't claim to understand the standards involved here, but maybe a feature idea would be to have a standard way to express to the DBAPI that this particular statement, as it is being executed, should enjoy enhanced logging or debugging output. Perhaps the use of a contextvar or other out-of-band signaling, or parametrization of the DBAPI calls could be an aspirational goal that allows programmers to get a more granular ability to interact with what the DBAPI is ultimately doing and what SQLA is ultimately emitting.

**Reply by CaselIT** [reply]
Depending on your code, you may turn on logging only in a single part of the system, like if using session with a engine as bind you could use
```py
session.bind.echo = True
... # your usual statements
session.bind.echo = False
```
Note that this will only work if the session is not in a transaction, since a connection will update its echo flag only on creation..

Another alternative to have fine grained control may also be to use an event, like [`before_cursor_execute`](https://docs.sqlalchemy.org/en/14/core/events.html#sqlalchemy.events.ConnectionEvents.before_cursor_execute) then pass a custom execution option to the sql you want to log:
```py
@listens_for(engine, "before_cursor_execute")
def go(conn, cursor, statement, parameters, context, executemany):
    if conn.get_execution_options().get("my_log"):
        print("log stuff", statement)


with engine.connect() as conn:
    conn.exec_driver_sql("select 1")
    with conn.execution_options(my_log=True) as c2:
        print(c2.exec_driver_sql("select 2").all())
```

**Reply by aathan** [reply]
It might be nice to collect some of these debugging tips and practices into a "howto"

The event mechanism may be more appropriate for this kind of debugging since, as I understand it, sessions are now "autobegin" and are therefore almost always already in a transaction. Then, the problem seems to shift to recognizing that this `statement` is the one you want to log. Perhaps a trick then would be to put a sentinel in the `parameters` or recognize the specific `parameters` you're looking for. Yes, these are the droids I'm ...

**Comment by zzzeek** [comment]
I will alter the FAQ entry to provide an alternate means of stringifying.

**Reply by zzzeek** [reply]
Here's the quickest way to stringify a statement with all params, for psycopg2, **insecurely, not for production use**.   The FAQ will have this example as well as other versions for positional paramstyles, mostly I have to catch enough dragons because there will be lots:

```py
import uuid

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import declarative_base


Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(UUID)

e = create_engine("postgresql+psycopg2://")


u1, u2, u3 = uuid.uuid4(), uuid.uuid4(), uuid.uuid4()
stmt = select(A).where(A.data.in_([u1, u2, u3]))

compiled = stmt.compile(e, compile_kwargs={"render_postcompile": True})

# assume psycopg2 pyformat style
# note this is NOT SECURE FOR PRODUCTION USE, NOT SECURE FOR UNTRUSTED INPUT,
# NOT SECURE AT ALL
print(str(compiled) % compiled.params)
```

**Reply by CaselIT** [reply]
> mostly I have to catch enough dragons because there will be lots:

let's hope they don't become extinct after that faq item :)

**Reply by zzzeek** [reply]
okey doke, back to the old ways of having two commits for everything, here's that:

b2a28c556f  
e760e0e1fe

**Reply by zzzeek** [reply]
here's the new section

https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#rendering-bound-parameters-inline

**Reply by aathan** [reply]
What is this fabled FAQ entry? (I was wrigint this as you posted the link above) Please still accept the feedback that this FAQ entry either did not exist when I was looking for it, or is too hard to find / should have references in the docs for the underlying system objects.

How can I find it and why didn't it come up when I searched and searched for how to get my hands on the SQL being sent to my DB engine?! Would you consider linking to it from the docs about "Rendering Bound Parameters" or at least mentioning `stmt.compile(... compile_kwargs={"render_postcompile": True})` in that area.

Please also don't lose sight of the fact that the documentation about TypeDecorators nowhere mentions how these are intended to be actually used. Unfortunately, if you are a sophisticated enough programmer, when you read those docs, you may start to think the type system is hooking these classes automatically somehow. So, merely mentioning that these decorators are for creating custom types that (crucially) you must then explicitly reference in your ORM declarations, would save a lot of head scratching. Likewise, mentioning that these can be included in a dialect specialization (per CaselIT's response above) AND how to interpose it (in `create_engine`?) so that the decorators will actually be used, would be a huge win for noobs like me.

**Reply by aathan** [reply]
By the way, wouldn't it make sense to make `WHERE a.data IN (%(data_1_1)s, %(data_1_2)s, %(data_1_3)s)` form the standard output of `str(sql)` or at least have an `interpolable_str()` that gives you something like that? Perhaps the `:data_1_1` format is some standard I'm forgetting about or not aware of?  It may look a little cleaner in logs but it sure isn't useful.

**Comment by CaselIT** [comment]
@aathan please understand that we don't like your condescending tone.
I think we provided more than enough help in good faith.

We will block the discussion if the same tone continues the get used..

**Reply by aathan** [reply]
My apologies. That was not my intent! I thought I was being funny, not condescending.

... and let me explicitly say:  Thank you for your responses, and a great framework!
